From 8f13a2ce7f2d192a898f090d25e8a8ce05912a97 Mon Sep 17 00:00:00 2001
From: Olivier Brunel <i.am.jack.mail@gmail.com>
Date: Sat, 13 Oct 2012 14:24:15 +0200
Subject: [PATCH 1/2] swap: fix swap behaviour with symlinks

Starting a swap unit pointing to (What) a symlink (e.g. /dev/mapper/swap
or /dev/disk/by-uuid/...) would have said unit marked active, following
the one using the "actual" device (/dev/{dm-1,sda3}), but that new unit
would be seen as inactive.
Since all requests to stop swap units would follow/redirect to it,
and it is seen inactive, nothing would be done (swapoff never called).

This is because this unit would be treated twice in
swap_process_new_swap, the second call to swap_add_one causing it to
eventually be marked inactive.

Signed-off-by: Olivier Brunel <i.am.jack.mail@gmail.com>
---
 src/core/swap.c | 10 +++++++---
 1 file changed, 7 insertions(+), 3 deletions(-)

diff --git a/src/core/swap.c b/src/core/swap.c
index c8e25d0..0184122 100644
--- a/src/core/swap.c
+++ b/src/core/swap.c
@@ -451,9 +451,13 @@ static int swap_process_new_swap(Manager *m, const char *device, int prio, bool
                                     st.st_rdev != udev_device_get_devnum(d))
                                         continue;
 
-                        k = swap_add_one(m, p, device, prio, false, false, set_flags);
-                        if (k < 0)
-                                r = k;
+                        /* Skip p==device, since that case will be handled below */
+                        if (!streq(p, device))
+                        {
+                            k = swap_add_one(m, p, device, prio, false, false, set_flags);
+                            if (k < 0)
+                                    r = k;
+                        }
                 }
 
                 udev_device_unref(d);
-- 
1.8.1.1

